pkg/model/mselection/heaviest.go

```diff
@@ -20,330 +18,303 @@ import (
-	tips := s.tipItems()
+// SelectTips tries to collect tips that confirm the most recent transactions since the last reset of the selector.
+// best tips are determined by counting the referenced transactions (heaviest branches) and by "removing" the
+// transactions of the referenced cone of the already choosen tips in the bitsets of the available tips.
+// only tips are considered that were present at the beginning of the SelectTips call,
+// to prevent attackers from creating heavier branches while we are searching the best tips.
+// "maxHeaviestBranchTipsPerCheckpoint" is the amount of tips that are collected if
+// the current best tip is not below "UnconfirmedTransactionsThreshold" before.
+// a minimum amount of selected tips can be enforced, even if none of the heaviest branches matches the
+// "minHeaviestBranchUnconfirmedTransactionsThreshold" criteria.
+// if at least one heaviest branch tip was found, "randomTipsPerCheckpoint" random tips are added
+// to add some additional randomness to prevent parasite chain attacks.
+// the selection is cancelled after a fixed deadline. in this case, it returns the current collected tips.
+func (s *HeaviestSelector) SelectTips(minRequiredTips int) (hornet.Hashes, error) {
+
+	// create a working list with the current tips to release the lock to allow faster iteration
+	// and to get a frozen view of the tangle, so an attacker can't
+	// create heavier branches while we are searching the best tips
+	// caution: the tips are not copied, do not mutate!
+	tipsList := s.tipsToList()
 
 	// tips could be empty after a reset
-	if len(tips) == 0 {
+	if tipsList.Len() == 0 {
 		return nil, ErrNoTipsAvailable
 	}
 
-	lastTip := tips[len(tips)-1]
+	var result hornet.Hashes
 
-	var best = struct {
-		tips  hornet.Hashes
-		count uint
-	}{
-		tips:  hornet.Hashes{lastTip.hash},
-		count: lastTip.refs.Count(),
-	}
+	// run the tip selection for at most 0.1s to keep the view on the tangle recent; this should be plenty
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(s.heaviestBranchSelectionDeadline))
+	defer cancel()
 
-	// loop through all tips and find the one with the most referenced transactions
-	for _, tip := range tips {
-		// when the context has been cancelled, return the current best with an error
+	deadlineExceeded := false
+
+	for i := 0; i < s.maxHeaviestBranchTipsPerCheckpoint; i++ {
+		// when the context has been cancelled, stop collecting heaviest branch tips
 		select {
 		case <-ctx.Done():
-			return randomTip(best.tips), ctx.Err()
+			deadlineExceeded = true
 		default:
 		}
 
-		c := tip.refs.Count()
-		if c > best.count {
-			best.tips = hornet.Hashes{tip.hash}
-			best.count = c
-		} else if c == best.count {
-			best.tips = append(best.tips, tip.hash)
+		tip, count, err := s.selectTip(tipsList)
+		if err != nil {
+			break
 		}
+
+		if (len(result) > minRequiredTips) && ((count < uint(s.minHeaviestBranchUnconfirmedTransactionsThreshold)) || deadlineExceeded) {
```

Comment:
This condition still confuses me
let say count is larger than `minHeaviestBranchUnconfirmedTransactionsThreshold`

why don't I break?

Reply:
We only break if it is lower, that's the whole point